src/contracts/core/DelegationManager.sol

```diff
@@ -535,137 +572,151 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
             );
         }
 
+        uint32 epochForEndOfSlashability = EpochUtils.getEndOfSlashabilityEpoch(withdrawalCreationEpoch(withdrawalRoot));
+        require(EpochUtils.currentEpoch() > epochForEndOfSlashability,
+            "DelegationManager._completeQueuedWithdrawal: withdrawal is still slashable");
+
         // Remove `withdrawalRoot` from pending roots
-        delete pendingWithdrawals[withdrawalRoot];
+        delete _pendingWithdrawalData[withdrawalRoot];
 
         // Finalize action by converting shares to tokens for each strategy, or
         // by re-awarding shares in each strategy.
-        if (receiveAsTokens) {
-            for (uint256 i = 0; i < withdrawal.strategies.length; ) {
-                require(
-                    withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
-                    "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
+        for (uint256 i = 0; i < withdrawal.strategies.length; ) {
+            require(
+                withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
+                "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
+            );
+
+            uint256 shares;
+            {
+                // Check if the withdrawal is withdrawable and get the scaling factor
+                (bool canWithdraw, uint64 scalingFactor) = slasher.getWithdrawabilityAndScalingFactorAtEpoch(
+                    withdrawal.delegatedTo,
+                    withdrawal.strategies[i],
+                    epochForEndOfSlashability
                 );
 
-                _withdrawSharesAsTokens({
-                    staker: withdrawal.staker,
-                    withdrawer: msg.sender,
-                    strategy: withdrawal.strategies[i],
-                    shares: withdrawal.shares[i],
-                    token: tokens[i]
+                require(canWithdraw, "DelegationManager._completeQueuedWithdrawal: withdrawal is still slashable");
+
+                shares = SlashingAccountingUtils.normalize({
+                    nonNormalizedShares: withdrawal.shares[i], 
+                    scalingFactor: scalingFactor
                 });
```

Comment:
This is fucking awful

After looking around, it looks like `normalize` is literally only called AFTER calls to the Slasher.

How about we have the Slashed do this normalization instead, and have it return already-normalized data?

Reply:
See #609